rules/firestore.json

```diff
@@ -35,50 +37,204 @@ service cloud.firestore {
+            // (typeGame, playerZero, playerOne, beginning) and performs safe updates:
+            // - lastUpdate must be updated monotonously by the corresponding player
+            // - turn can only be made +1 (usual turn), -1 (take back), or -2 (take back)
+            allow update: if userIsVerified() &&
+                             partHasStarted(resource.data) &&
+                             userIsPlayer(resource.data) &&
+                             preservesField('typeGame') &&
+                             preservesField('playerZero') &&
+                             preservesField('playerOne') &&
+                             preservesField('beginning') &&
+                             (preservesField('turn') || safeTurnUpdate(resource.data.turn, request.resource.data.turn)) &&
+                             (safeLastUpdate(resource.data.lastUpdate, request.resource.data.lastUpdate));
+            // All verified user can read parts (needed for the lobby)
+            allow read: if userIsVerified();
+        }
+        function userIsChosenOpponentInJoiner(partId) {
+            let joiner = get(/databases/$(database)/documents/joiners/$(partId)).data;
+            return hasField(joiner, 'chosenOpponent') && joiner.chosenOpponent.id == request.auth.uid;
+        }
+        function userIsPlayer(part) {
+            let userIsPlayerZero = part.playerZero.id == request.auth.uid;
+            let userIsPlayerOne = hasField(part, 'playerOne') && part.playerOne.id == request.auth.uid;
+            return userIsPlayerZero || userIsPlayerOne;
+        }
+        function safeTurnUpdate(oldTurn, newTurn) {
+            return (newTurn == oldTurn + 1) || (newTurn == oldTurn - 1) || (newTurn == oldTurn - 2);
+        }
+        function safeLastUpdate(oldLastUpdate, newLastUpdate) {
+            let legitimatePlayerZeroUpdate = newLastUpdate.player == 0 && resource.data.playerZero.id == request.auth.uid;
+            let legitimatePlayerOneUpdate = hasField(resource.data, 'playerOne') && newLastUpdate.player == 1 && resource.data.playerOne.id == request.auth.uid;
+            let indexIncremented = newLastUpdate.index == oldLastUpdate.index + 1;
+            return (legitimatePlayerZeroUpdate || legitimatePlayerOneUpdate || partHasStarted(resource.data) == false) && indexIncremented;
+        }
+        function partHasStarted(part) {
+            // A part is started when the beginning field is set
+            return hasField(part, 'beginning');
+        }
+        function hasTimedOut(user) {
+            let userDoc =  get(/databases/$(database)/documents/users/$(user.id)).data;
+            let timedOut = debug(request.time.time()) > debug(userDoc.last_changed.time()) + duration.value(10, 's');
+            return debug(timedOut);
+        }
+        match /joiners/{partId} {
+            // Any verified user can create a joiner as long as they are themselves creator and if the corresponding part exists
```

Comment:
and that a corresponding part exist :p ?

Reply:
sans le that, oui !